data/transactions/logic/jsonspec_test.go

```diff
@@ -0,0 +1,336 @@
+		require.NotNil(t, err)
+	}
+}
+
+func TestParseComments(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	text := `{"key0": /*comment*/"algo"}`
+	_, err := parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": "algo"}/*comment*/`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": [1,/*comment*/,3]}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+}
+
+func TestParseUnclosed(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	text := `{"key0": ["algo"}`
+	_, err := parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": ["algo"]]}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": ["algo"],"key1":{}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": ["algo"],"key1":{{}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": [1,}]}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+}
+
+func TestParseNested(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	text := `{"key0": [[1,2,3],[4,5,6]], "key1":{"key10":{"key100":"algo"}}}`
+	_, err := parseJSON([]byte(text))
+	require.Nil(t, err)
+}
+
+func TestParseWhiteSpace(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	//empty text
+	text := ""
+	_, err := parseJSON([]byte(text))
+	require.NotNil(t, err)
+	//space, tab, new line and carriage return are allowed
+	text = "{\"key0\": [\t]\n\r}"
+	_, err = parseJSON([]byte(text))
+	require.Nil(t, err)
+	//form feed is not allowed
+	text = "{\"key0\": [\f]}"
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+}
+
+func TestParseSpecialValues(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	text := `{"key0": NaN}`
+	_, err := parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": +Inf}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": -Inf}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": null}`
+	_, err = parseJSON([]byte(text))
+	require.Nil(t, err)
+}
+
+func TestParseHexValue(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	text := `{"key0": 0x1}`
+	_, err := parseJSON([]byte(text))
+	require.NotNil(t, err)
+	text = `{"key0": 0xFF}`
+	_, err = parseJSON([]byte(text))
+	require.NotNil(t, err)
+}
+
+func TestParseBigNum(t *testing.T) {
+	partitiontest.PartitionTest(t)
+	//max int can be represented using float64, 2<<52
+	text := `{"key0":9007199254740992}`
+	msg, err := parseJSON([]byte(text))
+	require.Nil(t, err)
+	require.True(t, json.Valid([]byte(text)))
+	require.Equal(t, uint64(9007199254740992), uint64(msg.(map[string]interface{})["key0"].(float64)))
+
+	//2<<52+1
+	text = `{"key0":9007199254740993}`
+	msg, err = parseJSON([]byte(text))
+	require.Nil(t, err)
+	require.True(t, json.Valid([]byte(text)))
+	require.NotEqual(t, uint64(9007199254740993), uint64(msg.(map[string]interface{})["key0"].(float64)))
+	require.Equal(t, uint64(9007199254740992), uint64(msg.(map[string]interface{})["key0"].(float64)))
```

Comment:
This is what we must avoid. The JSON _spec_ does not say that big integers get changed.  That's a common implementation issue, as many parsers convert the integers to floats.  I hope we can get Go not to do it, else we'll need a different parsing library.

(These are great tests, by the way, some of my earlier comments said we needed a lot of tests that I hadn't seen down here yet.)